internal/mesh/internal/controllers/explicitdestinations/controller.go

```diff
@@ -0,0 +1,235 @@
+		WithWatch(pbcatalog.WorkloadType, controller.ReplaceType(pbmesh.ComputedExplicitDestinationsType)).
+		WithReconciler(&reconciler{destinations: destinationsMapper})
+}
+
+type reconciler struct {
+	destinations *workloadselectionmapper.Mapper[*pbmesh.Destinations]
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, rt controller.Runtime, req controller.Request) error {
+	rt.Logger = rt.Logger.With("controller", ControllerName, "id", req.ID)
+
+	// Look up the associated workload.
+	workloadID := resource.ReplaceType(pbcatalog.WorkloadType, req.ID)
+	workload, err := resource.GetDecodedResource[*pbcatalog.Workload](ctx, rt.Client, workloadID)
+	if err != nil {
+		rt.Logger.Error("error fetching workload", "error", err)
+		return err
+	}
+
+	// If workload is not found, the decoded resource will be nil.
+	if workload == nil || workload.GetResource() == nil || workload.GetData() == nil {
+		// When workload is not there, we don't need to manually delete the resource
+		// because it is owned by the workload. In this case, we skip reconcile
+		// because there's nothing for us to do.
+		rt.Logger.Trace("the corresponding workload does not exist", "id", workloadID)
+		return nil
+	}
+
+	// Get existing ComputedDestinations resource (if any).
+	cpc, err := resource.GetDecodedResource[*pbmesh.ComputedExplicitDestinations](ctx, rt.Client, req.ID)
+	if err != nil {
+		rt.Logger.Error("error fetching ComputedDestinations", "error", err)
+		return err
+	}
+
+	// If workload is not on the mesh, we need to delete the resource and return
+	// as for non-mesh workloads there should be no destinations.
+	if !workload.GetData().IsMeshEnabled() {
+		rt.Logger.Trace("workload is not on the mesh, skipping reconcile and deleting any corresponding ComputedDestinations", "id", workloadID)
+
+		// Delete CD only if it exists.
+		if cpc != nil {
+			_, err = rt.Client.Delete(ctx, &pbresource.DeleteRequest{Id: req.ID})
+			if err != nil {
+				// If there's an error deleting CD, we want to re-trigger reconcile again.
+				rt.Logger.Error("error deleting ComputedDestinations", "error", err)
+				return err
+			}
+		}
+
+		// Otherwise, return as there's nothing else for us to do.
+		return nil
+	}
+
+	// Now get any destinations that we have in the cache that have selectors matching the name
+	// of this CD (name-aligned with workload).
+	destinationIDs := r.destinations.IDsForWorkload(req.ID)
+
+	decodedDestinations, err := r.fetchDestinations(ctx, rt.Client, destinationIDs)
+	if err != nil {
+		rt.Logger.Error("error fetching destinations", "error", err)
+		return err
+	}
+
+	newComputedDestinationsData := &pbmesh.ComputedExplicitDestinations{}
+	for _, dst := range decodedDestinations {
+		valid, cond := validate(ctx, rt.Client, dst)
+
+		// Only add it to computed destinations if its destinations are valid.
+		if valid {
+			newComputedDestinationsData.Destinations = append(newComputedDestinationsData.Destinations, dst.GetData().GetDestinations()...)
```

Comment:
Is it a problem if multiple `Destinations` resources add duplicates?

Reply:
Updated to find conflicts. I don't think we care about duplicate service references/ports, but if there is a listen address that is the same, it's good to update the status and avoid sending that data to envoy.